alphagov/pay-direct-debit-connector
# PP-4043 Add GoCardless Partner app state resource

## WHAT YOU DID

Context: 
- When linking a GoCardless merchant app to a GOV.UK Pay account using OAuth, we send a parameter called `state` to GoCardless. They will send this parameter back to us with the first GET request. We are going to use this query parameter as a check and validation between a gateway account that initiates the OAuth process and the access code that we receive from GoCardless.
- This work is required for `selfservice` to be able to create/get and disable this token against a gateway account.
- In the creation of the token, we check if there is any active token for this gateway account and if there's one we disable it first and then we create a new one. This functionality has been designed as a protection mechanism against requests with an old token that are coming in later.
- Selfservice will create a token when sending the GET request to GoCardless to get the access code, will verify the token against the gateway account id when it receives a request from GoCardless and will disable the token once the OAuth journey has finished.

- Added a resource to create, disable, get oauth tokens
- Added service for GoCardless partner app token
- Added Dao for GoCardless partner app token
- Added mocked resource test, that will assert the response code, headers
and the entity. I believe that spinning up a new HtpServlet environment and
a db are an overkill given that there are tests: the service in isolation,
the Dao with the database, the resource should be tested against Pact as a producer.
- Used a custom Dropwizard NonEmptyStringParam type, which treats empty
strings as missing. Also is easy to test a resource in isolation.

- Added resource where `selfservice` will send an `access code` from GoCardless Connect and `dd-connector` will make a POST request to GoCardless Connect to exchange the access code with an `access token`. After successfully getting the `access token` and `organisation id`, dd connector will update it internally and then reply to `selfservice`.

- Added a GoCardless Connect Client used to make a POST request to GoCardless Connect to exchange the access code with access token and organisation id.

## How to test

- How should it be reviewed?

- What feedback do you need?
Code style, functionality
- If manual testing is needed, give suggested testing steps.
Use Postman to call the resource.

## Code review checklist

### Logging
